server/src/core/tenant/datasource/services/datasource.service.ts

```diff
@@ -0,0 +1,210 @@
+import { ObjectMetadata } from 'src/core/tenant/datasource/entities/object-metadata';
+import { FieldMetadata } from 'src/core/tenant/datasource/entities/field-metadata.entity';
+import { baseColumns } from 'src/core/tenant/datasource/entities/base.entity';
+import {
+  convertFieldTypeToPostgresType,
+  sanitizeColumnName,
+  uuidToBase36,
+} from 'src/core/tenant/datasource/utils/datasource.util';
+
+@Injectable()
+export class DataSourceService implements OnModuleInit, OnModuleDestroy {
+  private mainDataSource: DataSource;
+  private metadataDataSource: DataSource;
+  private connectionOptions: PostgresConnectionOptions;
+  private dataSources = new Map<string, DataSource>();
+
+  constructor(environmentService: EnvironmentService) {
+    this.connectionOptions = {
+      url: environmentService.getPGDatabaseUrl(),
+      type: 'postgres',
+      logging: false,
+      schema: 'public',
+    };
+    this.mainDataSource = new DataSource(this.connectionOptions);
+    this.metadataDataSource = new DataSource({
+      ...this.connectionOptions,
+      schema: 'metadata',
+      synchronize: true, // TODO: remove this in production
+      entities: [DataSourceMetadata, ObjectMetadata, FieldMetadata],
+    });
+  }
+
+  /**
+   *
+   * Returns the schema name for a given workspaceId
+   * @param workspaceId
+   * @returns string
+   */
+  public getSchemaName(workspaceId: string): string {
+    return `workspace_${uuidToBase36(workspaceId)}`;
+  }
+
+  /**
+   * Creates a new schema for a given workspaceId
+   * @param workspaceId
+   * @returns Promise<void>
+   */
+  public async createWorkspaceSchema(workspaceId: string): Promise<void> {
+    const schemaName = this.getSchemaName(workspaceId);
+
+    const queryRunner = this.mainDataSource.createQueryRunner();
+    await queryRunner.createSchema(schemaName, true);
+    await queryRunner.release();
+
+    await this.insertNewDataSourceMetadata(workspaceId, schemaName);
+  }
+
+  /**
+   * Inserts a new data source
+   * @param workspaceId
+   * @param workspaceSchema this can be computed from the workspaceId but it won't be the case for remote data sources
+   * @returns Promise<void>
+   */
+  private async insertNewDataSourceMetadata(
+    workspaceId: string,
+    workspaceSchema: string,
+  ): Promise<void> {
+    await this.metadataDataSource
+      ?.createQueryBuilder()
+      .insert()
+      .into(DataSourceMetadata)
+      .values({
+        workspace_id: workspaceId,
+        schema: workspaceSchema,
+      })
+      .execute();
+  }
+
+  /**
+   * Connects to a workspace data source using the workspace metadata. Returns a cached connection if it exists.
+   * @param workspaceId
+   * @returns Promise<DataSource | undefined>
+   */
+  public async connectToWorkspaceDataSource(
+    workspaceId: string,
+  ): Promise<DataSource | undefined> {
+    // if (this.dataSources.has(workspaceId)) {
```

Comment:
Commented for now, each call creates a new connection (which is not destroyed automatically to my knowledge!), this is because I want to use synchronise mode for this draft. Will remove once we have proper migrations generated from metadata